src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,847 @@
+        private static readonly LocalizableString s_localizableUnsupportedOsMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PlatformCompatabilityCheckUnsupportedOsMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableUnsupportedOsVersionMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PlatformCompatabilityCheckUnsupportedOsVersionMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.PlatformCompatabilityCheckDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        // We are adding the new attributes into older versions of .Net 5.0, so there could be multiple referenced assemblies each with their own 
+        // version of internal attribute type which will cause ambiguity, to avoid that we are comparing the attributes by their name
+        private const string SupportedOSPlatformAttribute = nameof(SupportedOSPlatformAttribute);
+        private const string ObsoletedInOSPlatformAttribute = nameof(ObsoletedInOSPlatformAttribute);
+        private const string UnsupportedOSPlatformAttribute = nameof(UnsupportedOSPlatformAttribute);
+
+        // Platform guard method names
+        private const string IsOSPlatformVersionAtLeast = nameof(IsOSPlatformVersionAtLeast);
+        private const string IsOSPlatform = nameof(IsOSPlatform);
+        private const string IsBrowser = nameof(IsBrowser);
+        private const string IsLinux = nameof(IsLinux);
+        private const string IsFreeBSD = nameof(IsFreeBSD);
+        private const string IsFreeBSDVersionAtLeast = nameof(IsFreeBSDVersionAtLeast);
+        private const string IsAndroid = nameof(IsAndroid);
+        private const string IsAndroidVersionAtLeast = nameof(IsAndroidVersionAtLeast);
+        private const string IsIOS = nameof(IsIOS);
+        private const string IsIOSVersionAtLeast = nameof(IsIOSVersionAtLeast);
+        private const string IsMacOS = nameof(IsMacOS);
+        private const string IsMacOSVersionAtLeast = nameof(IsMacOSVersionAtLeast);
+        private const string IsTvOS = nameof(IsTvOS);
+        private const string IsTvOSVersionAtLeast = nameof(IsTvOSVersionAtLeast);
+        private const string IsWatchOS = nameof(IsWatchOS);
+        private const string IsWatchOSVersionAtLeast = nameof(IsWatchOSVersionAtLeast);
+        private const string IsWindows = nameof(IsWindows);
+        private const string IsWindowsVersionAtLeast = nameof(IsWindowsVersionAtLeast);
```

Comment:
Can we avoid the hard coded names for platform specific methods?  Basically `s_platformCheckMethodNames` will not be required. Instead `GetRuntimePlatformGuardMethods` can be adjusted so it filters method name by regex `IsXXXVersionAtLeast`.

Reply:
As I mentioned before personally would prefer `ImmutableDictionary` over `Regex`, but I would let our leads decide on this
CC @bartonjs @terrajobst @jeffhandley 

Note that i am ok with `IsXXXVersionAtLeast` Regex, but `IsXXX` Regex might match different purpose APIs added to the `OperatingSystem` class in the future (even with additional return type parameter check)

Reply:
I think we'd have to search for `OperatingSystem.Is(\w*)`, right? Otherwise you wouldn't catch the version-less guard methods. It could very well pick up false positive matches in the future, but we're going to have to be careful on the `OperatingSystem` type to never add an `IsXxx` function that isn't a platform.

We should support the regex approach for extensibility, but if there's a perf gain for using an `ImmutableDictionary` for the known platforms and then falling back to the regex, that seems OK to me.

Reply:
> I think we'd have to search for OperatingSystem.Is(\w*), right?

We will be filtering from `OperatingSystem` type methods so no need to check for the type OperatingSystem, as i said i am ok with regex-ing `IsXXXVersionAtLeast` and we might add versioning for browser in the future, but `IsXXX` could give false positives in the future (there is no such method other than OS check for now)

If time allows i could try both and compare the benchmarking results

Reply:
I must be missing something, but wouldn't `IsXxx()` be needed to match versionless guards like `IsWindows()`?  I think it's OK to include `IsXxx`. If we add another `OperatingSystem.IsXxx()` function in the future, we can update the analyzer at the same time.

Reply:
> I must be missing something, but wouldn't IsXxx() be needed to match versionless guards like IsWindows()

Yes it is, what i meant is it can give false positives in the future

> If we add another OperatingSystem.IsXxx() function in the future, we can update the analyzer at the same time.

If it is not OS check method and we are still aware of need to update the analyzer. Anyway if you guys OK with that the fine, i can update accordingly

Reply:
OK, thanks.  I appreciate the forward-compatibility caution, but I since we also control the `OperatingSystem` type, we'll just have to transfer that caution to the API reviews when modifying the `OperatingSystem` type in the future.

But to balance it, we could also check (if not already checked) that it's a `static` function that returns `bool`.

Reply:
We can ensure no false positives if it is guaranteed that `IsXxxVersionAtLeast()` and `IsXxx()` always ship in pairs. The analyzer logic will then first compute all `IsXxxVersionAtLeast()` methods and only include `IsXxx()` methods if there is a corresponding `IsXxxVersionAtLeast()` already identified. If the paired shipping is not guaranteed, hardcoding is fine if false positives is a concern.

Reply:
`IsBrowser()` does not have an `IsBrowserVersionAtLeast()` pair since it's an unversioned platform. Same with Linux.

Reply:
My recommendation is hard-coded until we see what growth looks like.

Reply:
> My recommendation is hard-coded until we see what growth looks like.

OK. I'd be comfortable with that too.

Reply:
Sounds good, thanks!